[aarch64] Add Sbgemm kernel to accelerate fp32 tensor matmul with bfloat16

snadampal opened the pull request:
### Description
<!-- Describe your changes. -->
This PR adds SbgemmKernel for aarch64. This includes Sbegmm kernel to implement matrix multiplication with bfloat16 SIMD instructions (bfmmla) and MatMul operator changes to invoke the Sbgemm kernel. To enable Sbgemm kernel, set the following session option:
"kOrtSessionOptionsGemmFastMathMode"

The PR also adds new test cases for mlas and ort.

### Motivation and Context
<!-- - Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here. -->
This is to improve MatMul performance on aarch64 platform.
I have run the below benchmarking script (bert , roberta and gpt2 model inference) on AWS Graviton3 based c7g.4xl instance and observed 1.2x -1.76x performance improvement compared to sgemm (fp32) kernel performance.

```
cd onnxruntime/python/tools/transformers
python3 benchmark.py
```
And the unit test precision results are matching to sgemm kernel results.
`./build.sh --config RelWithDebInfo --build_shared_lib --parallel --compile_no_warning_as_error --skip_submodule_sync `

snadampal commented:
appreciate if someone can review this PR.

snadampal commented:
Hi, I have rebased the PR to resolve the merge conflicts. I'm happy to address any feedback you may have. Thank you!

milpuz01 commented:
I have checked out the changes and run performance test and accuracy tests with and without flag using `onnxruntime_perf_test` (modified the binary to dump output for comparisons) on AWS Graviton3 instances and it was fine.

yufenglee commented on onnxruntime/core/providers/cpu/math/matmul.h:
```diff
@@ -15,6 +15,14 @@ class MatMul final : public OpKernel {
   Status Compute(OpKernelContext* context) const override;
 };
 
+static inline bool ort_use_fastmath_mode() {
```
could you please replace the env with a session option?

snadampal replied:
thanks, @yufenglee , will replace.

yufenglee commented on onnxruntime/core/providers/cpu/math/matmul.h:
```diff
@@ -48,6 +58,12 @@ class MatMul<float> final : public OpKernel {
   bool trans_batch_b_;
+
+  //fastmath mode state
+  bool use_fastmath_mode;
```
nit: use_fastmath_mode_.

yufenglee commented on onnxruntime/core/providers/cpu/math/matmul.h:
```diff
@@ -48,6 +58,12 @@ class MatMul<float> final : public OpKernel {
+  bool use_fastmath_mode;
+  // sbgemm kernel is implemented as 8x8 blocks with weights pre-packed to 4 blocks of 4x2
+  // so a minimum of 32 elements is defined to outweigh the additional prepacking overhead
+  const size_t fastmath_mode_kernelsize_threshold = 32;
```
nit: kFastMathModeKernelsizeThreshold

chenfucn commented on onnxruntime/core/providers/cpu/math/matmul.h:
```diff
@@ -15,6 +15,14 @@ class MatMul final : public OpKernel {
   Status Compute(OpKernelContext* context) const override;
 };
 
+static inline bool ort_use_fastmath_mode() {
```
For compile time feature switch, use a compilation flag. For runtime feature switch, use a session option.

https://github.com/microsoft/onnxruntime/blob/main/include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h

snadampal replied:
Thanks, @chenfucn , I will take a loot at the session options.

yufenglee commented on onnxruntime/core/providers/cpu/math/matmul.cc:
```diff
@@ -126,15 +126,64 @@ Status MatMul<T>::Compute(OpKernelContext* ctx) const {
   return Status::OK();
 }
 
+bool GemmPackBbf16(AllocatorPtr& alloc,
```
nit: GemmPackBBfloat16

snadampal commented:
Hi @chenfucn , @yufenglee , I have updated the PR (1) to move to the newer gemm interface and (2) to add session option based fastmath mode control. Please review and let me know your feedback.

snadampal commented:
Hi @chenfucn , @yufengle, appreciate if someone can trigger the CI for this PR. I have addressed all the feedback except the windows testing for which I'm waiting for the Windows CI results. Thank you!

chenfucn commented on onnxruntime/core/providers/cpu/math/matmul.cc:
```diff
@@ -187,21 +243,37 @@ Status MatMul<float>::Compute(OpKernelContext* ctx) const {
+      data[i].alpha = alpha_attr_;
+      data[i].beta = 0.0f;
   }
   MlasGemmBatch(trans_a ? CblasTrans : CblasNoTrans, trans_b ? CblasTrans : CblasNoTrans,
```
Is the indentation correct? Looks like the fp32 gemm is called outside of the if then else branch?

chenfucn commented on onnxruntime/test/providers/base_tester.cc:
```diff
@@ -41,6 +41,14 @@ void DebugTrap() {
 
+static inline bool use_cosine_similarity() {
+  static auto value = [&] {
+    const char* ptr = std::getenv("ORT_TEST_USE_COSINE_SIMILARITY");
```
I am not sure about test environment setup, but historically ORT does not use environment variables as switches. Adding this would make the testing pipeline more complex, causing either flaky tests or missed unit testing.

snadampal replied:
removed all the env variables and used the the existing arguments to decide on cosine similarity checks.

chenfucn reviewed (commented):
As we discussed, please add mlas unit tests that call the kernel directly with different shapes are other parameters.

chenfucn commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux OpenVINO CI Pipeline, Linux QNN CI Pipeline, MacOS CI Pipeline

snadampal commented on onnxruntime/core/mlas/lib/sbgemm.h:
```diff
@@ -0,0 +1,539 @@
+    // Compute the number of bytes required to hold the packed buffer.
+    //
+    const auto* dispatch = MlasSBGemmGetDispatch();
+    const auto padding = dispatch->BufOverRead;
```
thanks. given there is no generic sbgemm, yes, the dispatcher can be null on other devices. I will fix it.

snadampal commented on cmake/onnxruntime_mlas.cmake:
```diff
@@ -339,15 +339,21 @@ else()
+          set_source_files_properties(${MLAS_SRC_DIR}/sbgemm_kernel_neon.cpp PROPERTIES COMPILE_FLAGS " -march=armv8.2-a+bf16 ")
+          set_source_files_properties(${MLAS_SRC_DIR}/aarch64/SbgemmKernelNeon.S PROPERTIES COMPILE_FLAGS " -march=armv8.2-a+bf16 ")
+
+          set(CMAKE_ASM_FLAGS "${CMAKE_ASM_FLAGS} -march=armv8.2-a+bf16")
```
Hi @yufenglee, sorry, I didn't understand the comment. My understanding is we build the whole project for "generic" arch and selectively enable the isa extensions wherever required.

snadampal replied:
i got it and fixed it in the latest version.

snadampal commented:
Thanks for the review, I will update the PR to address this and also add unit tests.

snadampal commented:
I have updated the PR to address all the feedback so far and also the learnings from my other qgemm PR. (1) added mlas unit tests (2) tested linux full build (both release and release with debug info) (3) minimal build (4) android build with cross compilation on x86. and (5)  lintrunner and git-clang-format

Next, adding ort optimizer and provider tests to test the fastmath session. 
Please review and let me know if any feedback on this  version.

snadampal commented:
I have added the matmul fastmath tests for ort optimizer and EP math op.

chenfucn commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux OpenVINO CI Pipeline, Linux QNN CI Pipeline, MacOS CI Pipeline

chenfucn commented:
/azp run ONNX Runtime Web CI Pipeline, Windows ARM64 QNN CI Pipeline, Windows CPU CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, onnxruntime-binary-size-checks-ci-pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed

snadampal commented:
I have fixed the new sbgemm test failures on x86 builds.

snadampal commented:
Hi @chenfucn , could you please launch only x86 builds/tests in CI selectively, once they pass, we can start the rest. I have locally tested aarch64 linux, x86 linux and aarch64 android builds.

snadampal commented:
Hi @chenfucn , @yufenglee , Could one of you please start the CI runs? Thank you!

yufenglee commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux OpenVINO CI Pipeline, Linux QNN CI Pipeline, MacOS CI Pipeline

yufenglee commented:
/azp run ONNX Runtime Web CI Pipeline, Windows ARM64 QNN CI Pipeline, Windows CPU CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, onnxruntime-binary-size-checks-ci-pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed

chenfucn commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux OpenVINO CI Pipeline, Linux QNN CI Pipeline, MacOS CI Pipeline

chenfucn commented:
/azp run ONNX Runtime Web CI Pipeline, Windows ARM64 QNN CI Pipeline, Windows CPU CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, onnxruntime-binary-size-checks-ci-pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed

snadampal commented:
Hi @chenfucn , the CI failures are coming from the newly added sbgemm mlas tests, duplicate definitions , but i'm not able to reproduce it with the following builds. Any documentation on how to repro the CI builds ?
```
onnxruntime_mlas_test
./build.sh --config RelWithDebInfo --build_shared_lib --parallel --compile_no_warning_as_error --skip_submodule_sync
./build.sh --config=Release --build_shared_lib --build_wheel --parallel 16
```

snadampal commented:
I have fixed the CI failures for cross compilation builds. Could someone please trigger the CI runs?
thanks to @snnn for helping with the CI build instructions.

snnn commented on onnxruntime/test/util/compare_ortvalue.cc:
```diff
@@ -65,6 +66,53 @@ const char* ElementTypeToString(MLDataType type) {
+  std::pair<COMPARE_RESULT, std::string> res = std::make_pair(COMPARE_RESULT::SUCCESS, "");
+  const float cosine_similarity_threshold = 0.99;
+
+  T dot = 0.0, denom_a = 0.0, denom_b = 0.0;
```
Add 'f' to each of the constants. Like, 0.0f. Without of the suffix the literals would be in double type instead of float. Float values can be implicitly converted to double, but not vice versa.  Or you may add a static_cast here.

snadampal replied:
thanks, @snnn . I have fixed and pushed a new version.

snadampal commented:
Hi, can someone please trigger the CI if there are no more comments on the PR. Thank you!

snadampal commented:
multiple OSes, devices and compiler combinations is making the PR more challenging. So, I have taken more conservative approach and enabled the feature only for aarch64/linux, updated the PR. I'm wondering if only a subset of CI (probably windows builds alone) can be run to avoid redundant runs.

snnn commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux OpenVINO CI Pipeline, Linux QNN CI Pipeline, MacOS CI Pipeline

snnn commented:
/azp run ONNX Runtime Web CI Pipeline, Windows ARM64 QNN CI Pipeline, Windows CPU CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, onnxruntime-binary-size-checks-ci-pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed

snadampal commented:
there are 3 failures:
1. ##[section]Starting: 7. Extended minimal build with NNAPI EP for Android(arm64-v8a) and skip tests.
logs: [link](https://dev.azure.com/onnxruntime/2a773b67-e88b-4c7f-9fc0-87d31fea8ef2/_apis/build/builds/1244317/logs/29)
error: 
```
In file included from /onnxruntime_src/onnxruntime/core/mlas/lib/tanh.cpp:23:
2023-12-12T02:55:34.4317995Z /onnxruntime_src/onnxruntime/core/mlas/lib/mlasi.h:372:11: error: unknown type name 'bfloat16_t'; did you mean 'float16_t'?
```

I have tried the following build on x86 Linux platform but not able to reproduce the above error. **Can someone please tell me how to repro this?**

` --build_dir /build/7 --cmake_generator Ninja --config MinSizeRel --skip_submodule_sync --parallel --android --android_sdk_path /android_home --android_ndk_path /ndk_home --android_abi=arm64-v8a --android_api=29 --use_nnapi --minimal_build extended --build_shared_lib --disable_ml_ops --disable_exceptions --skip_tests `

2. unreferenced parameter treated as error. I will fix it
`warning C4100: 'enable': unreferenced formal parameter`

3. tensorRT and training test failures. these are not related to the the PR

snadampal commented:
Looks like the difference between my setup where bfloat16_t is recognized vs the failing CI setup is the clang version. When I downloaded the android ndk, I got Clang17, where as the CI failure log shows Clang14. Is there any specific Android NDK version that I need to test?
cc: @snnn @skottmckay @chenfucn

snnn commented:
All our pipelines should use Android NDK 26.1.10909125. But the "Linux CPU Minimal Build E2E CI Pipeline" is not. I will upgrade it now.

snnn commented:
Our macOS pipelines use AppleClang 14.  They probably will hit the same problem.

snadampal commented:
I made the feature aarch64 && linux only, so I believe we dont see this on Mac.

snadampal commented:
Hi @snnn , I've fixed the unused parameter error, expecting the clang version issue to be addressed with the CI upgrade.

snadampal commented:
Hi @snnn , I see the Linux CPU min build CI is upgraded,  can you please check if this PR is good now?

snnn commented:
/azp run Windows GPU TensorRT CI Pipeline, Windows x64 QNN CI Pipeline, onnxruntime-binary-size-checks-ci-pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed

snnn commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux OpenVINO CI Pipeline, Linux QNN CI Pipeline, MacOS CI Pipeline, Windows ARM64 QNN CI Pipeline, Windows CPU CI Pipeline, Windows GPU CI Pipeline

snnn commented on include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h:
```diff
@@ -235,3 +235,9 @@ static const char* const kOrtSessionOptionsOptimizedModelExternalInitializersFil
+// Option values:
+// - "0": Gemm FastMath mode is not enabled. [DEFAULT]
+// - "1": Gemm FastMath mode is enabled.
+static const char* const kOrtSessionOptionsGemmFastMathMode = "session.enable_gemm_fastmath_mode";
```
This is confusing. What is fastmath? 
First, complier also has a fast math mode. Whether it was enabled or not, at runtime we cannot change it.
Second, we have multiple gemm implementations. Most of them do not support this switch. If we add an option here but it has no effect, users may report it as a bug.

snadampal replied:
Hi @snnn ,
(1) fastmath mode will enable SBGemm kernel which takes fp32 weigths, activations and perform gemm in Bfloat16 precision to accelerate. I have used the similar terminology as in Arm compute Library [ACL](https://github.com/ARM-software/ComputeLibrary).
(2) yes, when disabled,these kernels are not even invoked. the regular SGemm kernels get used and produce the same result for the same input
(3) this is not the compiler fast math mode. this can be enabled by setting the session options at runtime. I have added test cases for this,
   ```
 SessionOptions so;
    ASSERT_STATUS_OK(so.config_options.AddConfigEntry(
        kOrtSessionOptionsGemmFastMathMode, "1"));
```
(4) In the commit and the PR I stated it is only for aarch64 and sgemm acceleration with sbgemm only. But please let me know if I need to rename the flag or update any doc to state the scope explicitly.

snnn commented:
You need to run clang-format on some of the files you changed. 

https://github.com/microsoft/onnxruntime/actions/runs/7202794995/job/19767804736?pr=17031

skottmckay commented on include/onnxruntime/core/framework/op_kernel.h:
```diff
@@ -129,6 +129,12 @@ class OpKernel {
     return Status::OK();
   }
 
+  // Override this function to handle the fastmath implementation mode of the kernel
+  // the default behavior is ignore this request
+  virtual Status SetFastMathMode(bool /*enable = true*/) {
+    return Status::OK();
+  }
```
I think this is way too specific to be adding to the base OpKernel class, and the virtual method is going to impact binary size. We need a more generic way to make these extremely specific config values available to the kernel. 

One option would be to add the SessionOptions to the CPU EP (we do this for the [XNNPACK EP](https://github.com/microsoft/onnxruntime/blob/6d7519ede8298a422e84e70bfdf01cc46fbf76c3/onnxruntime/core/providers/xnnpack/xnnpack_provider_factory.cc#L13)). The EP instance should be accessible via OpKernelInfo.GetExecutionProvider() in the MatMul ctor, so if CPUExecutionProvider exposes a way to read config values the kernel can access them. 

We can look for other options as well.

skottmckay replied:
Alternatively adding a helper to OpKernelInfo that allows reading a config option might be slightly cleaner/simpler. We will need to discuss and agree upon the best approach which may require waiting until some folks that are currently on vacation return.

e.g.

```c++
class OpKernelInfo : public OpNodeProtoHelper<ProtoHelperNodeContext> {
 public:
  explicit OpKernelInfo(const onnxruntime::Node& node,
                        const KernelDef& kernel_def,
                        const IExecutionProvider& execution_provider,
                        const std::unordered_map<int, OrtValue>& constant_initialized_tensors,
                        const OrtValueNameIdxMap& mlvalue_name_idx_map,
                        const DataTransferManager& data_transfer_mgr,
                        const ConfigOptions& config_options,
                        const AllocatorMap& allocators = {});
  ...
  // call GetConfigEntry instead of having a basic getter for ConfigOptions so we don't need to add the entire ConfigOptions class to the provider bridge
  std::optional<std::string> GetConfigOption(const std::string& key) const {
    return config_options_.GetConfigEntry(key);
  }
```

snadampal commented:
Hi @skottmckay , thanks for the review. Some of the other existing session options were just for the graph rewrite stage, so they ended at the optimizer class and didn't had to be extended beyond, but for this option, since it is required all the way to kernel compute time, I took the OpKernel base class approach. If there is any other simpler way, I'm happy to rework the PR.  And yes, will update the doc and config name to reflect that it's only for aarch64 mlas gemm.

Hi @snnn , regarding clang-format, when i tried to run it last time as part of another PR, it had updated the whole mlas project, I think mlas files formatting is different than clang. I will try again only specific to the new files and see.

snadampal commented:
I will update this PR once the session config support [PR](https://github.com/microsoft/onnxruntime/pull/18897) is merged.

snadampal commented:
I have reworked the fastmath session control logic to use the newly added config options via OpKernelInfo. Please let me know if the PR looks good now. thank you!
@skottmckay , @snnn , @chenfucn

snnn commented:
/azp run Windows CPU CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, Windows x64 QNN CI Pipeline, onnxruntime-binary-size-checks-ci-pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed

snnn commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux OpenVINO CI Pipeline, Linux QNN CI Pipeline, MacOS CI Pipeline, Windows ARM64 QNN CI Pipeline

snadampal commented:
of the 4 failures, 2 are lint tests, and the rest 2 are not related to this PR.

Windows GPU CI pipeline and Windows GPU CI Pipeline (training build_x64_RelWithDebInfo)  failure is cuda related.
```
======================================================================
ERROR: test_arena_with_cuda_graph (__main__.TestInferenceSessionWithCudaGraph)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "onnxruntime_test_python_cudagraph.py", line 155, in test_arena_with_cuda_graph
    session.run_with_iobinding(io_binding)
  File "D:\a\_work\1\b\RelWithDebInfo\RelWithDebInfo\onnxruntime\capi\onnxruntime_inference_collection.py", line 331, in run_with_iobinding
    self._sess.run_with_iobinding(iobinding._iobinding, run_options)
RuntimeError: Error in execution: D:\a\_work\1\s\onnxruntime\core\providers\cuda\cuda_call.cc:121 onnxruntime::CudaCall D:\a\_work\1\s\onnxruntime\core\providers\cuda\cuda_call.cc:114 onnxruntime::CudaCall CUDA failure 901: operation failed due to a previous error during capture ; GPU=0 ; hostname=a04a8a05c000001 ; file=D:\a\_work\1\s\onnxruntime\core\providers\cuda\cuda_graph.cc ; line=33 ; expr=cudaStreamEndCapture(stream_, &graph_); 
```

snnn commented:
@snadampal , have you tried to run clang-format on these files that were listed in the lint errors?

snadampal commented:
When I tried the clang-format, it was changing the whole file, i will go through the error lines and try to fix one by one.
one more thing I have to change is the session option name and add it to the docs.
Is the following session name acceptable?

`static const char* const kOrtSessionOptionsMlasGemmFastMathMode = "session.enable_mlas_gemm_fastmath_mode";
`

Please let me know the docs to be updated.

snnn commented:
Here is all the format change needed:

https://github.com/microsoft/onnxruntime/commit/211eb92eaa7175bed9ad31a1e13d4049c35b0523

You may cherry-pick it to your branch.

snadampal commented:
thanks for formatting the code, @snnn, I have cherry-picked it.

Hi @skottmckay , if you could comment on the naming part, I will update the PR. the following is what I think should cover your suggestion. You had suggested replacing session with mlas. I think session makes sense because the config is still session specific, so I left it, but added mlas to it.

`static const char* const kOrtSessionOptionsMlasGemmFastMathMode = "session.enable_mlas_gemm_fastmath_mode"; 
`

snnn commented:
@snadampal  did you push your change to Github?

snadampal commented:
not yet, would like to push the code format changes along with the session name change

snadampal commented:
Hi @skottmckay , appreciate your response on this.

> Hi @skottmckay , if you could comment on the naming part, I will update the PR. the following is what I think should cover your suggestion. You had suggested replacing session with mlas. I think session makes sense because the config is still session specific, so I left it, but added mlas to it.
> 
> `static const char* const kOrtSessionOptionsMlasGemmFastMathMode = "session.enable_mlas_gemm_fastmath_mode"; `

skottmckay commented:
> Hi @skottmckay Scott McKay, appreciate your response on this.
> 
> > Hi @skottmckay
> >             
> >             Scott McKay
> >             , if you could comment on the naming part, I will update the PR. the following is what I think should cover your suggestion. You had suggested replacing session with mlas. I think session makes sense because the config is still session specific, so I left it, but added mlas to it.
> > `static const char* const kOrtSessionOptionsMlasGemmFastMathMode = "session.enable_mlas_gemm_fastmath_mode"; `

I think I would consider the first name as something that points me to where I would find the setting being used. e.g. 'optimization' means look in the optimizer project. I would say it's inferred you're configuring something in the session as you're using SessionOptions (vs. say RunOptions). Based on that, I would vote for 'mlas.' as the prefix.

The name also seems a little too generic as it sounds like it would apply to MLAS as a whole. Unless we think there will be some other fastpath that applies to MLAS GEMM in general, a more specific name would be clearer. e.g. `mlas.enable_gemm_fastmath_arm64_bfloat16` 

Or alternatively the platform/datatype could be in the value and you could parse that. 

e.g. `mlas.enable_gemm_fastmath_mode` could have a value of `arm64.bfloat16` and additional platform.data_type values could be parsed for. That is obviously more complicated so we should avoid unless we think it would be used. Possibly required if we had this GEMM fastpath on multiple platforms or for multiple data types though as I assume I'd want to be able to enable/disable each specific combination, and having a new config key for every single combination doesn't scale well.

yufenglee commented on onnxruntime/core/mlas/lib/qgemm_kernel_smmla.cpp:
```diff
@@ -121,7 +121,7 @@ MlasGemmQuantCopyPackA<MLAS_GEMM_S8S8_KERNEL_SMMLA>(
         const int8_t* a7 = a0 + lda * 7;
 
-        size_t k = CountK;
+        int k = static_cast<int>(CountK);
```
>int k = static_cast<int>(CountK); [](http://example.com/codeflow?start=8&length=33)

nit: why this change?

yufenglee commented on onnxruntime/core/mlas/lib/qgemm_kernel_ummla.cpp:
```diff
@@ -122,7 +122,7 @@ MlasGemmQuantCopyPackA<MLAS_GEMM_U8X8_KERNEL_UMMLA>(MLAS_GEMM_U8X8_KERNEL_UMMLA:
         const uint8_t* a7 = a0 + lda * 7;
 
-        size_t k = CountK;
+        int k = static_cast<int>(CountK);
```
same here. we don't need this change.

snadampal replied:
this was the warning from android build I think. the file doesn't belong to sbgemm kernels. I will put a separate PR for this.

yufenglee commented on onnxruntime/core/mlas/lib/sbgemm.h:
```diff
@@ -0,0 +1,445 @@
+
+#include <cassert>
+#include <cstdlib>
+#include <string>
```
Do you need this head file?

snadampal replied:
will double check

snadampal replied:
fixed.

snadampal commented:
thank you, I see your point. bf16 and f16 are the potential fastmath options, but on aarch64, so far I see interest for bf16 fastmath alone. I agree that there may not be multiple of these for different platforms, so I will go ahead with a simple config key.

`static const char* const kOrtSessionOptionsMlasGemmFastMathArm64Bfloat16 = "mlas.enable_gemm_fastmath_arm64_bfloat16"; `

chenfucn commented on onnxruntime/core/providers/cpu/math/matmul.cc:
```diff
@@ -134,7 +172,24 @@ Status MatMul<float>::PrePack(const Tensor& tensor, int input_idx, /*out*/ Alloc
+
+    if (use_fastmath_mode_ && (trans_b_attr_ == 0) && ((dim1 * dim2) >= kFastMathModeKernelsizeThreshold)) {
+      is_packed = GemmPackBBfloat16(alloc, tensor, trans_b_attr_ != 0, packed_b_, packed_b_size, b_shape_);
+    } else
```
what if we are on an ARM64 chip that does not support bf16? Then the GemmPackBBfloat16 should fail. But then we should fall back to packing fp32. This logic does not seem right to me.

snadampal replied:
`use_fastmath_mode_` is set to true only for the platforms that have bf16 support and the mode is explicitly enabled through session options.

```
#if defined(__aarch64__) && defined(__linux__)
    auto config_ops = info.GetConfigOptions().GetConfigEntry(kOrtSessionOptionsGemmFastMathMode);
    use_fastmath_mode_ = (config_ops == "1") && CPUIDInfo::GetCPUIDInfo().HasArmNeon_BF16();
#endif
```

chenfucn commented on onnxruntime/test/mlas/unittest/test_sbgemm.cpp:
```diff
@@ -0,0 +1,141 @@
+}
+
+static UNUSED_VARIABLE bool added_to_main = AddTestRegister([](bool is_short_execute) {
+  if (!MlasBf16AccelerationSupported()) {
```
sorry but I can't find this function called from production code? only test code? What if the system is run on an arm chip with no bf16 support? Can you point me to the code that detect cpu feature and then decide which kernel to dispatch?

snadampal replied:
I have implemented the fastmath mode checks in two places.
(1) inside CPU provider matmul: Here I'm using the ortsession option && cpuifo id to decide whether bf16 kernel can be dispatched. anything going through ORT is covered with this check

(2) inside sbgemm kernel: This is the `MlasBf16AccelerationSupported` which I've implemented using cpuid check, and i'm using it for Mlas tests which don't go through ort.

chenfucn replied:
why do you need this MlasBf16AccelerationSupported function in this case?

snadampal replied:
anyone using Mlas directly. for now, while running Mlas unit tests on aach64 + linux platform, this check guards those don't have bf16 support.
probably I can avoid cpuinfo checks inside matmul.h and use this Mlas function instead, will update it.

chenfucn replied:
yeah it looks strange that an interface function is only used in test code.

chenfucn commented on onnxruntime/core/mlas/lib/platform.cpp:
```diff
@@ -70,6 +82,10 @@ MLASCPUIDInfo::MLASCPUIDInfo()
     has_arm_sve_i8mm_ = ((getauxval(AT_HWCAP2) & HWCAP2_SVEI8MM) != 0);
+
+    has_arm_sve_ = ((getauxval(AT_HWCAP) & HWCAP_SVE) != 0);
+    has_arm_neon_bf16_ = ((getauxval(AT_HWCAP2) & HWCAP2_BF16) != 0);
```
how are these feature flag used? it seems to me that they are only used in test code? could you point me to the logic that the feature flag is used to determine which kernel to call during runtime?

snadampal replied:
these flags are used in the following logic to detect whether bf16 feature is supported or not.

1. cpu provider matmul.h
https://github.com/microsoft/onnxruntime/pull/17031/files#diff-b64379f9113c5d88446bf978474921915149b31ca703038f06673b81cf552a48R33

```
#if defined(__aarch64__) && defined(__linux__)
    auto config_ops = info.GetConfigOptions().GetConfigEntry(kOrtSessionOptionsGemmFastMathMode);
    use_fastmath_mode_ = (config_ops == "1") && CPUIDInfo::GetCPUIDInfo().HasArmNeon_BF16();
#endif
```

2. and in sbgemm 
https://github.com/microsoft/onnxruntime/pull/17031/files#diff-a6732e6798dee7a36040e9c388882279bbb70f1e53372a0a751b149429346118

```
bool MLASCALL
MlasBf16AccelerationSupported()
{
#if defined(MLAS_TARGET_ARM64)
    return MLAS_CPUIDINFO::GetCPUIDInfo().HasArmNeon_BF16();
#else
    return false;
#endif
```

chenfucn replied:
What about sve_bf16? is it useful? Should it be HasArmNeon_BF16() || HasArmSVE_BF16() instead?  If here the feature we care about is bf16 support, does it make sense to provide a simpler interface like HasBf16()?

snadampal replied:
Currently I have added only neon bf16 kernels, not SVE. so, i'm not using sve_bf16() flags yet. but defined them for future use.
regarding unifying them to single interface, it's a good suggestion. but given they are two different extensions and can be supported independently , I think it's better to check for each of them separately and implement kernels accordingly .

chenfucn replied:
Are neon-bf16 and sve-bf16 require different kernels? If that is the case, then maybe you need more changes to have kernel function pointer setup based on hardware feature detection like:
https://github.com/microsoft/onnxruntime/blob/f45ef1da184c3c74e8e935cdaf12b67dcd953ddb/onnxruntime/core/mlas/lib/platform.cpp#L540 ?

It would be great if you can add comments accordingly to explain purpose of this flag and future steps to use it. or just remove it.

snadampal replied:
thanks. I will remove the SVE flags. I wouldn't expect a scenario where both neon and sve kernels are active at the same time. since SVE is a superset of scenarios neon covers, I would assume people will prefer SVE over neon and deprecate the older ones. so, the kernel function pointer will always be one, pointing to the highest performant one.

snadampal commented:
Update the PR for the session name and other points discussed so far including the clang-formatting. Tested
1. release, debug and minimal builds on aarch64 neoverse v1 and n1 platforms
2. android build and linux cross compilation for aarch64 config on x86 platform

chenfucn commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux OpenVINO CI Pipeline, Linux QNN CI Pipeline, MacOS CI Pipeline, Windows ARM64 QNN CI Pipeline

chenfucn commented:
/azp run Windows CPU CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, Windows x64 QNN CI Pipeline, onnxruntime-binary-size-checks-ci-pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed

snadampal commented:
Thanks to @chenfucn , @snnn , @skottmckay and @yufenglee for the great feedback and merging the PR!

snnn commented:
@snadampal , thanks for making ONNX Runtime better. Welcome to bring more changes to us.  You have my email. Do not hesitate to contact me anytime when you need help on reviewing PRs.

maajidkhann commented:
Hello @snadampal . This is a great reference PR for any SIMD based contributions to be made for ARM.
Can you please help me with how do we generate this file (onnxruntime/core/mlas/lib/aarch64/**SbgemmKernelNeon.S**)?
https://github.com/microsoft/onnxruntime/pull/17031/files#diff-6458cefb29cdb4ba0a976ca7ba93e0f3738f6b02e8d6063a51378c4fecfba7c4

My understanding is, we can add the SIMD intrinsic in the .cpp file (onnxruntime/core/mlas/lib/**sbgemm_kernel_neon.cpp**) https://github.com/microsoft/onnxruntime/pull/17031/files#diff-a6732e6798dee7a36040e9c388882279bbb70f1e53372a0a751b149429346118 like how you have added NEON code here and then use the gcc/clang compiler to generate the .S or .asm file?

snadampal commented:
Hi @maajidkhann , Intrinsics based will be the best approach because it scales well for new architectures. but for this PR, I had hand written the assembly.  since the goal to extract the best performance.

snnn commented:
This PR has been cherry-picked into the `rel-1.17.0` branch in PR #19243. Removing the `release:1.17.0` label.
